arc.m

```diff
@@ -855,11 +869,6 @@ OBJC_PUBLIC id objc_loadWeakRetained(id* addr)
 	// will acquire the lock before attempting to deallocate)
 	if (obj == nil)
 	{
-		// If we've destroyed this weak ref, then make sure that we also deallocate the object.
```

Comment:
This change shouldn't matter for functional correctness, but you are removing an optimisation (i.e. dropping the refcount of the weak ref struct once we notice that it's no longer needed).

That said, looking at this code I believe that this is probably the real source of the bug that you're seeing.  We're unconditionally decrementing the refcount of we weakref structure, but we're then only removing our reference if it is the last weak reference.